nano/node/node.cpp

```diff
@@ -1340,6 +1340,252 @@ bool nano::node::init_error () const
+			std::greater<uint64_t>>,
+		boost::multi_index::hashed_unique<boost::multi_index::tag<account_tag>,
+			boost::multi_index::member<account_upgrade_item, nano::account, &account_upgrade_item::account>>>>
+	accounts_list;
+	// clang-format on
+
+	bool finished_upgrade (false);
+
+	while (!finished_upgrade && !stopped)
+	{
+		bool finished_accounts (false);
+		uint64_t total_upgraded_accounts (0);
+		while (!finished_accounts && count_limit != 0 && !stopped)
+		{
+			{
+				auto transaction (store.tx_begin_read ());
+				// Collect accounts to upgrade
+				for (auto i (store.latest_begin (transaction)), n (store.latest_end ()); i != n; ++i)
+				{
+					nano::account const & account (i->first);
+					nano::account_info const & info (i->second);
+					if (info.epoch () < epoch_a)
+					{
+						release_assert (nano::epochs::is_sequential (info.epoch (), epoch_a));
+						accounts_list.emplace (account_upgrade_item{ account, info.modified });
+					}
+				}
+			}
+
+			/* Upgrade accounts
+			Repeat until accounts with previous epoch exist in latest table */
+			std::atomic<uint64_t> upgraded_accounts (0);
+			std::atomic<uint64_t> workers (0);
+			for (auto i (accounts_list.get<modified_tag> ().begin ()), n (accounts_list.get<modified_tag> ().end ()); i != n && upgraded_accounts < upgrade_batch_size && upgraded_accounts < count_limit && !stopped; ++i)
+			{
+				std::atomic<bool> to_next_account (false);
+				auto transaction (store.tx_begin_read ());
+				nano::account_info info;
+				nano::account const & account (i->account);
+				if (!store.account_get (transaction, account, info) && info.epoch () < epoch_a)
+				{
+					auto difficulty (nano::work_threshold (nano::work_version::work_1, nano::block_details (epoch_a, false, false, true)));
+					nano::root const & root (info.head);
+					std::shared_ptr<nano::block> epoch = builder.state ()
+					                                     .account (account)
+					                                     .previous (info.head)
+					                                     .representative (info.representative)
+					                                     .balance (info.balance)
+					                                     .link (link)
+					                                     .sign (raw_key, signer)
+					                                     .work (0)
+					                                     .build ();
+					if (threads != 0)
+					{
+						++workers;
+						while (workers > threads)
+						{
+							nano::unique_lock<std::mutex> lock (mutex);
+							condition.wait (lock);
+						}
+						worker.push_task ([node_l = shared_from_this (), &upgrader_process, &condition, &upgraded_accounts, &workers, &to_next_account, epoch, difficulty, signer, root, account]() {
+							upgrader_process (*node_l, upgraded_accounts, to_next_account, epoch, difficulty, signer, root, account);
+							--workers;
+							condition.notify_all ();
```

Comment:
I forgot to add this. Every modification of `workers` must be done under the `mutex` to prevent a deadlock. But in this case it might just be easier to do change the wait conditions to `wait_for`.

Reply:
If we go the protection by mutex route, then `workers` could be non-atomic

Reply:
Thanks!